psydac/linalg/tests/test_linalg.py

```diff
@@ -33,6 +31,47 @@ def compute_global_starts_ends(domain_decomposition, npts):
 
     return global_starts, global_ends
 
+def get_StencilVectorSpace(n1, n2, p1, p2, P1, P2):
+    npts = [n1, n2]
+    pads = [p1, p2]
+    periods = [P1, P2]
+    D = DomainDecomposition(npts, periods=periods)
+    global_starts, global_ends = compute_global_starts_ends(D, npts)
+    C = CartDecomposition(D, npts, global_starts, global_ends, pads=pads, shifts=[1,1])
+    V = StencilVectorSpace(C)
+    return V
+
+def get_positive_definite_nparray(n1, n2):
+    np.random.seed(2)
+    mat = np.zeros((n1*n2, n1*n2))
+    for i in range(1, n1):
+        R = np.random.random((n2, n2))
+        for j in range(n1-i):
+            mat[j*n2:(j+1)*n2, (i+j)*n2:(i+j+1)*n2] = R
+    mat += np.transpose(mat)
+    sums = [sum(mat[i, :]) for i in range(n1*n2)]
+    max_sums = [max([sums[i*n2+j] for i in range(n1)]) for j in range(n2)]
+    D = np.random.random((n2, n2))
+    D += np.transpose(D)
+    for i in range(n2):
+        z = [D[i, (i+j)%n2] for j in range(1, n2)]
+        D[i, i] *= (max_sums[i] + sum(z)) / ( D[i, i] * np.random.random(1) )
+    for i in range(n1):
+        mat[i*n2:(i+1)*n2, i*n2:(i+1)*n2] = D
+    mat /= n1*n2 # to avoid exploding determinants
+    return mat
+
+def nparray_to_stencil(mat, V, n1, n2):
+    S = StencilMatrix(V, V)
+    for k in range(n1):
+        for l in range(0-k, n1-k):
+            loc_mat = np.reshape(mat[k*n2:(k+1)*n2, (l+k)*n2:(l+k+1)*n2], n2**2)
+            indices = [[k,i,l,j] for i in range(n2) for j in range(0-i, n2-i)]
+            for n, idx in enumerate(indices):
+                S[idx] = loc_mat[n]
+    S.remove_spurious_entries()
+    return S
```

Comment:
I use this method only for matrices created with `get_positive_definite_nparray`, i.e., matrices of the right specific tensor-product banded structure that you mentioned above. However, this method actually fills every matrix entry individually and hence requires no structure whatsoever.

Suggestions: I keep `get_positive_definite_nparray` the way it is, but write `nparray_to_stencil` in such a way that it will always output a StencilMatrix of the right specific tensor-product banded structure?

Reply:
That is automatically true. A `StencilMatrix` object only stores the non-zero elements of a matrix which has tensor-product banded structure. 😉

Reply:
7fc1cb2dfbbaad9fab59690f2a18f850489f4d30